apps/users/src/app/services/suggestion.service.v2.ts

```diff
@@ -0,0 +1,139 @@
+ * under the terms of the GNU General Public License version 3 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * version 3 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 3 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Castcle, 22 Phet Kasem 47/2 Alley, Bang Khae, Bangkok,
+ * Thailand 10160, or visit www.castcle.com if you need additional information
+ * or have any questions.
+ */
+
+import {
+  AuthenticationServiceV2,
+  DataService,
+  UserServiceV2,
+} from '@castcle-api/database';
+import { Meta, PaginationQuery, UserField } from '@castcle-api/database/dtos';
+import { User } from '@castcle-api/database/schemas';
+import { Authorizer } from '@castcle-api/utils/decorators';
+import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
+import { Cache } from 'cache-manager';
+
+@Injectable()
+export class SuggestionServiceV2 {
+  constructor(
+    @Inject(CACHE_MANAGER) private cacheManager: Cache,
+    private dataService: DataService,
+    private authenticationServiceV2: AuthenticationServiceV2,
+    private userServiceV2: UserServiceV2,
+  ) {}
+
+  async suggest(authorizer: Authorizer, pageQuery: PaginationQuery) {
+    let sliceUsers: User[] = [];
+    let users: User[] = [];
+    if (pageQuery.sinceId || pageQuery.untilId) {
+      users = await this.querySuggestByCache(authorizer, pageQuery);
+      let currentUserIndex = users.findIndex(
+        (user) => user._id === pageQuery.sinceId,
+      );
+      if (pageQuery.untilId) {
+        currentUserIndex++;
+        sliceUsers = users.slice(
+          currentUserIndex,
+          currentUserIndex + pageQuery.maxResults,
+        );
+      } else {
+        sliceUsers = users.slice(
+          Math.max(currentUserIndex - pageQuery.maxResults, 0),
+          pageQuery.maxResults,
+        );
+      }
+    } else {
+      users = await this.querySuggestByDataScience(authorizer);
+      sliceUsers = users.slice(0, pageQuery.maxResults);
+    }
+
+    const usersConvert = await this.userServiceV2.convertUsersToUserResponsesV2(
+      authorizer.user,
+      sliceUsers,
+      pageQuery.userFields?.includes(UserField.Relationships),
+    );
+
+    return {
+      payload: usersConvert,
+      meta: Meta.fromDocuments(sliceUsers),
+    };
+  }
+
+  private async querySuggestByCache(
+    authorizer: Authorizer,
+    pageQuery: PaginationQuery,
+  ) {
+    let suggestUser: { engagements: number; userId: string }[] =
+      await this.cacheManager.get(
+        this._suggestKey(authorizer.credential.accessToken),
```

Comment:
```suggestion
        CacheKey.ofSuggesUser(authorizer.credential.accessToken),
```
Should create a new class for storing all cache key patterns